internal/gcsx/random_reader.go

```diff
@@ -159,10 +184,68 @@ func (rr *randomReader) CheckInvariants() {
 	}
 }
 
+// tryReadingFromFileCache creates the cache handle first if it doesn't exist already
+// and then use that handle to read object's content which is cached in local file.
+// For the successful read, it returns number of bytes read, and a boolean representing
+// cacheHit as true.
+// For unsuccessful read, returns cacheHit as false, in this case content
+// should be read from GCS.
+// And it returns non-nil error in case something unexpected happens during the execution.
+// In this case, we must abort the Read operation.
+func (rr *randomReader) tryReadingFromFileCache(ctx context.Context,
+	p []byte,
+	offset int64) (n int, cacheHit bool, err error) {
+	if rr.fileCacheHandler != nil {
+		// Create fileCacheHandle if not already.
+		if rr.fileCacheHandle == nil {
+			rr.fileCacheHandle, err = rr.fileCacheHandler.GetCacheHandle(rr.object, rr.bucket, rr.downloadForRandomRead, offset)
+			if err != nil {
+				return 0, false, fmt.Errorf("tryReadingFromFileCache: while creating CacheHandle instance: %v", err)
+			}
+		}
+
+		n, err = rr.fileCacheHandle.Read(ctx, rr.object, offset, p)
+		if err == nil {
+			cacheHit = true
+			return
+		}
+
+		cacheHit = false
+		n = 0
+		if util.IsCacheHandleInvalid(err) {
+			err = rr.fileCacheHandle.Close()
+			if err != nil {
+				err = fmt.Errorf("ReadAt: while closing the fileCacheHandle: %v", err)
+				return
+			}
+			rr.fileCacheHandle = nil
+		} else if !strings.Contains(err.Error(), util.FallbackToGCSErrMsg) {
+			err = fmt.Errorf("ReadAt: while reading via cache: %v", err)
+			return
+		}
+		err = nil
+	}
+	return
+}
+
 func (rr *randomReader) ReadAt(
 	ctx context.Context,
 	p []byte,
-	offset int64) (n int, err error) {
+	offset int64) (n int, cacheHit bool, err error) {
+
+	n, cacheHit, err = rr.tryReadingFromFileCache(ctx, p, offset)
+	if err != nil {
+		err = fmt.Errorf("ReadAt: while reading from cache: %v", err)
+		return
+	}
+	if cacheHit {
+		return
+	}
+
+	// Otherwise fallback to read from GCS.
+	cacheHit = false
```

Comment:
nit: is it required to reset these vars ?

Reply:
No, removing this.